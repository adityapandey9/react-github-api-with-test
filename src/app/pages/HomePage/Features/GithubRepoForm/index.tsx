import React from 'react';
import styled from 'styled-components/macro';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { FormLabel } from 'app/components/FormLabel';
import { Input } from './components/Input';
import { TextButton } from './components/TextButton';
import { selectUsername } from './slice/selectors';
import { useGithubRepoFormSlice } from './slice';
import { REPO_LIST_PATH } from '../../../../path';
import { messages } from '../../messages';
import { useTranslation } from 'react-i18next';

export function GithubRepoForm() {
  const { t } = useTranslation();

  const { actions } = useGithubRepoFormSlice();

  const username = useSelector(selectUsername);

  const history = useHistory();

  const dispatch = useDispatch();

  const onChangeUsername = (evt: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(
      actions.changeUsername(evt.currentTarget.value.replace(/\s+/, '')),
    );
  };

  const onSubmitForm = (evt?: React.FormEvent<HTMLFormElement>) => {
    /* istanbul ignore next  */
    if (evt !== undefined && evt.preventDefault && username.length > 0) {
      evt.preventDefault();
      history.push(REPO_LIST_PATH.replace(':userName', username));
    }
  };

  return (
    <Wrapper>
      <FormGroup onSubmit={onSubmitForm}>
        <InputWrapper>
          <Input
            type="text"
            placeholder={t(messages.searchBoxPlaceholder())}
            value={username}
            onChange={onChangeUsername}
          />
        </InputWrapper>
        <UiButton type="submit">{t(messages.searchBoxBtnTxt())}</UiButton>
      </FormGroup>
    </Wrapper>
  );
}

const Wrapper = styled.div`
  ${TextButton} {
    margin: 16px 0;
    font-size: 0.875rem;
  }
`;

const InputWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;

  ${Input} {
    width: ${100 / 2}%;
    margin-right: 0.5rem;
  }
`;

const FormGroup = styled.form`
  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;

  ${FormLabel} {
    margin-bottom: 0.25rem;
    margin-left: 0.125rem;
  }
`;

const UiButton = styled.button`
  display: flex;
  align-self: center;
  text-align: center;
  background: ${p => p.theme.background};
  // outline: none;
  padding: 10px 60px;
  margin: 20px 0px;
  border: 1px solid rgb(220, 120, 95);
  border-radius: 10px;
  color: ${p => p.theme.primary};
  cursor: pointer;

  &:hover {
    opacity: 0.8;
  }

  &:active {
    opacity: 0.4;
  }
`;
