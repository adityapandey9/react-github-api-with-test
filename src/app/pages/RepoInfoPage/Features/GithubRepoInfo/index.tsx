import React from 'react';
import styled from 'styled-components/macro';
import { useSelector, useDispatch } from 'react-redux';
import ReactMarkdown from 'react-markdown';

import { TextButton } from './components/TextButton';
import {
  selectGithubRepoInfoError,
  selectGithubRepoInfoRepos,
  selectGithubRepoInfoLoading,
} from './slice/selectors';
import { LoadingIndicator } from 'app/components/LoadingIndicator';
import { RepoErrorType } from './slice/types';
import { useGithubRepoInfoSlice } from './slice';
import { useRepoIdChange } from './slice/hook';
import { RepoItem } from './RepoItem';
import { useUserNameChange } from '../../../ReposListPage/Features/GithubRepoList/slice/hook';
import { selectUsername } from '../../../HomePage/Features/GithubRepoForm/slice/selectors';
import { selectRepoListRepoId } from '../../../ReposListPage/Features/GithubRepoList/slice/selectors';

export function GithubRepoInfo() {
  const { actions } = useGithubRepoInfoSlice();
  useUserNameChange();
  useRepoIdChange();

  const username = useSelector(selectUsername);
  const repoId = useSelector(selectRepoListRepoId);
  const repoInfo = useSelector(selectGithubRepoInfoRepos);
  const isLoading = useSelector(selectGithubRepoInfoLoading);
  const error = useSelector(selectGithubRepoInfoError);

  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(actions.loadRepoById());
  }, [username, repoId, dispatch, actions]);

  return (
    <Wrapper>
      {isLoading && <LoadingIndicator small />}
      {repoInfo?.tree && repoInfo?.tree.length > 0 ? (
        <InfoContainer>
          <List>
            {repoInfo.tree.map(repo => (
              <RepoItem
                key={repo.sha}
                name={repo.path}
                fileSize={repo.size}
                url={repo.url}
              />
            ))}
          </List>
          <ReactMarkdown className="markDown">
            {repoInfo.readMe !== '' && repoInfo.readMe
              ? repoInfo.readMe
              : '## Sorry, no README.md file exists.'}
          </ReactMarkdown>
        </InfoContainer>
      ) : error ? (
        <ErrorText>{repoErrorText(error)}</ErrorText>
      ) : null}
    </Wrapper>
  );
}

export const repoErrorText = (error: RepoErrorType) => {
  switch (error) {
    case RepoErrorType.USER_NOT_FOUND:
      return 'There is no such repo ðŸ˜ž';
    case RepoErrorType.USERNAME_EMPTY:
      return 'Type any Github repo or username';
    case RepoErrorType.USER_HAS_NO_REPO:
      return 'User has no repository ðŸ¥º';
    case RepoErrorType.GITHUB_RATE_LIMIT:
      return 'Looks like github api`s rate limit(60 request/h) has exceeded ðŸ¤”';
    default:
      return 'An error has occurred!';
  }
};

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 60vh;

  ${TextButton} {
    margin: 16px 0;
    font-size: 0.875rem;
  }
  .markDown {
    color: ${p => p.theme.text} !important;
    padding: 15px;
    margin-top: 15px;
    border: 2px solid ${p => p.theme.border};
    border-radius: 8px;

    & > pre {
      background: ${p => p.theme.text};
      color: ${p => p.theme.backgroundVariant};
    }
  }
`;

const ErrorText = styled.span`
  color: ${p => p.theme.text};
`;

const List = styled.div``;

const InfoContainer = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
`;
