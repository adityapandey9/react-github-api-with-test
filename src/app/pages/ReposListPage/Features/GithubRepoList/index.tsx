import React from 'react';
import styled from 'styled-components/macro';
import { useSelector, useDispatch } from 'react-redux';
import { RepoItem } from './RepoItem';
import { TextButton } from './components/TextButton';
import {
  selectRepoListError,
  selectRepoListLoading,
  selectRepoListRepos,
} from './slice/selectors';
import { LoadingIndicator } from 'app/components/LoadingIndicator';
import { RepoErrorType } from './slice/types';
import { useGithubRepoListSlice } from './slice';
import { selectUsername } from '../../../HomePage/Features/GithubRepoForm/slice/selectors';
import { useUserNameChange } from './slice/hook';
import { useHistory } from 'react-router-dom';
import { REPO_INFO_PATH } from '../../../../path';

export function GithubRepoList() {
  const { actions } = useGithubRepoListSlice();
  useUserNameChange();

  const username = useSelector(selectUsername);
  const repos = useSelector(selectRepoListRepos);
  const isLoading = useSelector(selectRepoListLoading);
  const error = useSelector(selectRepoListError);

  const dispatch = useDispatch();
  const history = useHistory();

  React.useEffect(() => {
    dispatch(actions.loadRepos());
  }, [username, dispatch, actions]);

  return (
    <Wrapper>
      {isLoading && <LoadingIndicator small />}
      {repos?.length > 0 ? (
        <List>
          {repos.map(repo => (
            <RepoItem
              key={repo.id}
              onClick={event => {
                event.preventDefault();
                history.push(
                  REPO_INFO_PATH.replace(':userName', username).replace(
                    ':repoId',
                    repo.name,
                  ),
                );
              }}
              name={repo.name}
              starCount={repo.stargazers_count}
              url={repo.html_url}
            />
          ))}
        </List>
      ) : error ? (
        <ErrorText>{repoErrorText(error)}</ErrorText>
      ) : null}
    </Wrapper>
  );
}

export const repoErrorText = (error: RepoErrorType) => {
  switch (error) {
    case RepoErrorType.USER_NOT_FOUND:
      return 'There is no such user ðŸ˜ž';
    case RepoErrorType.USERNAME_EMPTY:
      return 'Type any Github username';
    case RepoErrorType.USER_HAS_NO_REPO:
      return 'User has no repository ðŸ¥º';
    case RepoErrorType.GITHUB_RATE_LIMIT:
      return 'Looks like github api`s rate limit(60 request/h) has exceeded ðŸ¤”';
    default:
      return 'An error has occurred!';
  }
};

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 60vh;

  ${TextButton} {
    margin: 16px 0;
    font-size: 0.875rem;
  }
`;

const ErrorText = styled.span`
  color: ${p => p.theme.text};
`;

const List = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
`;
