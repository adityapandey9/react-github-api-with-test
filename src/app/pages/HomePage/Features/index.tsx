import * as React from 'react';
import { GithubRepoForm } from './GithubRepoForm';

export function Features() {
  return (
    <>
      <GithubRepoForm />
    </>
  );
}
