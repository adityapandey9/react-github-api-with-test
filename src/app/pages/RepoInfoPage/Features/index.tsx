import * as React from 'react';
import { GithubRepoInfo } from './GithubRepoInfo';
import { Lead } from '../../ReposListPage/Features/GithubRepoList/components/Lead';
import { useSelector } from 'react-redux';
import { selectUsername } from '../../HomePage/Features/GithubRepoForm/slice/selectors';
import { selectRepoListRepoId } from '../../ReposListPage/Features/GithubRepoList/slice/selectors';
import { useTranslation } from 'react-i18next';
import { messages } from '../messages';

export function Features() {
  const { t } = useTranslation();

  const username = useSelector(selectUsername);
  const repoId = useSelector(selectRepoListRepoId);

  return (
    <>
      <Lead>
        {t(messages.getTitle())}{' '}
        <strong>
          {username}
          {'/'}
          {repoId}
        </strong>
      </Lead>
      <GithubRepoInfo />
    </>
  );
}
