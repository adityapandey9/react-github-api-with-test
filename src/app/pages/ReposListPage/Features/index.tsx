import * as React from 'react';
import { GithubRepoList } from './GithubRepoList';
import { Lead } from './GithubRepoList/components/Lead';
import { useSelector } from 'react-redux';
import { selectUsername } from '../../HomePage/Features/GithubRepoForm/slice/selectors';
import { useTranslation } from 'react-i18next';
import { messages } from '../messages';

export function Features() {
  const { t } = useTranslation();

  const username = useSelector(selectUsername);

  return (
    <>
      <Lead>
        {t(messages.getTitle())} <strong>{username}</strong>
      </Lead>
      <GithubRepoList />
    </>
  );
}
