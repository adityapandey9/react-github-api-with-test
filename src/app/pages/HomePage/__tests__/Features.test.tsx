import * as React from 'react';
import { Features } from '../Features';
import i18next from 'i18next';
import { Store } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { ThemeProvider } from '../../../../styles/theme/ThemeProvider';
import { HelmetProvider } from 'react-helmet-async';
import { BrowserRouter, Switch } from 'react-router-dom';
import { configureAppStore } from '../../../../store/configureStore';
import TestRenderer from 'react-test-renderer';

const { act } = TestRenderer;

const renderFeature = (store: Store) =>
  TestRenderer.create(
    <Provider store={store}>
      <ThemeProvider>
        <HelmetProvider>
          <BrowserRouter>
            <Switch>
              <Features />
            </Switch>
          </BrowserRouter>
        </HelmetProvider>
      </ThemeProvider>
    </Provider>,
  );

describe('HomePage <Features />', () => {
  let store: ReturnType<typeof configureAppStore>;
  let component: ReturnType<typeof renderFeature>;

  beforeEach(() => {
    store = configureAppStore();
    component = renderFeature(store);
  });
  afterEach(() => {
    component.unmount();
  });

  it('should render with en translations', () => {
    component = renderFeature(store);
    act(() => {
      /* fire events that update state */
      i18next.changeLanguage('en');
    });
    expect(component).toMatchSnapshot();
  });

  it('should render with de translations', () => {
    component = renderFeature(store);
    act(() => {
      /* fire events that update state */
      i18next.changeLanguage('de');
    });
    expect(component).toMatchSnapshot();
  });

  it('should render different after language change', () => {
    const renderedOutput1 = renderFeature(store);
    act(() => {
      /* fire events that update state */
      i18next.changeLanguage('de');
    });

    const renderedOutput2 = renderFeature(store);
    act(() => {
      /* fire events that update state */
      i18next.changeLanguage('en');
    });

    expect(renderedOutput1).not.toEqual(renderedOutput2);
  });
});
