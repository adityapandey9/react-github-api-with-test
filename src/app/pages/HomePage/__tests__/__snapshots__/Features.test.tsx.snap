// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage <Features /> should render with de translations 1`] = `
.c4 {
  border: 2px solid rgba(58,52,51,0.12);
  border-radius: 4px;
  color: rgba(58,52,51,1);
  display: block;
  width: 100%;
  font-size: 1rem;
  font-weight: 500;
  line-height: 1.5;
  color: rgba(58,52,51,1);
  background-color: rgba(255,255,255,1);
  -webkit-transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  outline: none;
  height: 2.75rem;
  padding: 0 0.75rem;
}

.c4::-webkit-input-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4::-moz-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4:-ms-input-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4::placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4:focus {
  border-color: rgba(215,113,88,1);
  box-shadow: 0 0 0 3px rgba(215,113,88,0.2);
}

.c0 .TextButton-sc-18yutv9-0 {
  margin: 16px 0;
  font-size: 0.875rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 .c3 {
  width: 50%;
  margin-right: 0.5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
}

.c1 .FormLabel-sc-1aa0nfe-0 {
  margin-bottom: 0.25rem;
  margin-left: 0.125rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  text-align: center;
  background: rgba(255,255,255,1);
  padding: 10px 60px;
  margin: 20px 0px;
  border: 1px solid rgb(220,120,95);
  border-radius: 10px;
  color: rgba(215,113,88,1);
  cursor: pointer;
}

.c5:hover {
  opacity: 0.8;
}

.c5:active {
  opacity: 0.4;
}

<div
  className="c0"
>
  <form
    className="c1"
    onSubmit={[Function]}
  >
    <div
      className="c2"
    >
      <input
        className="c3 c4"
        onChange={[Function]}
        placeholder="Type any Github username"
        type="text"
        value="adityapandey9"
      />
    </div>
    <button
      className="c5"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;

exports[`HomePage <Features /> should render with en translations 1`] = `
.c4 {
  border: 2px solid rgba(58,52,51,0.12);
  border-radius: 4px;
  color: rgba(58,52,51,1);
  display: block;
  width: 100%;
  font-size: 1rem;
  font-weight: 500;
  line-height: 1.5;
  color: rgba(58,52,51,1);
  background-color: rgba(255,255,255,1);
  -webkit-transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  outline: none;
  height: 2.75rem;
  padding: 0 0.75rem;
}

.c4::-webkit-input-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4::-moz-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4:-ms-input-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4::placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c4:focus {
  border-color: rgba(215,113,88,1);
  box-shadow: 0 0 0 3px rgba(215,113,88,0.2);
}

.c0 .TextButton-sc-18yutv9-0 {
  margin: 16px 0;
  font-size: 0.875rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 .c3 {
  width: 50%;
  margin-right: 0.5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
}

.c1 .FormLabel-sc-1aa0nfe-0 {
  margin-bottom: 0.25rem;
  margin-left: 0.125rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  text-align: center;
  background: rgba(255,255,255,1);
  padding: 10px 60px;
  margin: 20px 0px;
  border: 1px solid rgb(220,120,95);
  border-radius: 10px;
  color: rgba(215,113,88,1);
  cursor: pointer;
}

.c5:hover {
  opacity: 0.8;
}

.c5:active {
  opacity: 0.4;
}

<div
  className="c0"
>
  <form
    className="c1"
    onSubmit={[Function]}
  >
    <div
      className="c2"
    >
      <input
        className="c3 c4"
        onChange={[Function]}
        placeholder="Type any Github username"
        type="text"
        value="adityapandey9"
      />
    </div>
    <button
      className="c5"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
